process/block/preprocess/transactions.go

```diff
@@ -920,256 +924,142 @@ func (txs *transactions) createAndProcessMiniBlocksFromMeV1(
-
-	mapGasConsumedByMiniBlockInReceiverShard := make(map[uint32]uint64)
+	log.Debug("createAndProcessMiniBlocksFromMeXX has been started")
 
-	gasInfo := gasConsumedInfo{
-		gasConsumedByMiniBlockInReceiverShard: uint64(0),
-		gasConsumedByMiniBlocksInSenderShard:  uint64(0),
-		totalGasConsumedInSelfShard:           txs.gasHandler.TotalGasConsumed(),
+	args := miniBlocksBuilderArgs{
+		gasTracker:                txs.gasTracker,
+		accounts:                  txs.accounts,
+		accountTxsShards:          &txs.accountTxsShards,
+		balanceComputationHandler: txs.balanceComputation,
+		blockSizeComputation:      txs.blockSizeComputation,
+		haveTime:                  haveTime,
+		isShardStuck:              isShardStuck,
+		isMaxBlockSizeReached:     isMaxBlockSizeReached,
+		getTxMaxTotalCost:         getTxMaxTotalCost,
 	}
 
-	log.Debug("createAndProcessMiniBlocksFromMeV1", "totalGasConsumedInSelfShard", gasInfo.totalGasConsumedInSelfShard)
-
-	senderAddressToSkip := []byte("")
+	mbBuilder, err := newMiniBlockBuilder(args)
+	if err != nil {
+		return nil, err
+	}
 
 	defer func() {
 		go txs.notifyTransactionProviderIfNeeded()
 	}()
 
-	for shardID := uint32(0); shardID < txs.shardCoordinator.NumberOfShards(); shardID++ {
-		mapMiniBlocks[shardID] = txs.createEmptyMiniBlock(txs.shardCoordinator.SelfId(), shardID, block.TxBlock, nil)
-	}
-
-	mapMiniBlocks[core.MetachainShardId] = txs.createEmptyMiniBlock(txs.shardCoordinator.SelfId(), core.MetachainShardId, block.TxBlock, nil)
-
-	for index := range sortedTxs {
-		if !haveTime() {
-			log.Debug("time is out in createAndProcessMiniBlocksFromMeV1")
-			break
-		}
-
-		tx, ok := sortedTxs[index].Tx.(*transaction.Transaction)
-		if !ok {
-			log.Debug("wrong type assertion",
-				"hash", sortedTxs[index].TxHash,
-				"sender shard", sortedTxs[index].SenderShardID,
-				"receiver shard", sortedTxs[index].ReceiverShardID)
-			continue
-		}
-
-		txHash := sortedTxs[index].TxHash
-		senderShardID := sortedTxs[index].SenderShardID
-		receiverShardID := sortedTxs[index].ReceiverShardID
-
-		miniBlock, ok := mapMiniBlocks[receiverShardID]
-		if !ok {
-			log.Debug("mini block is not created", "shard", receiverShardID)
-			continue
-		}
-
-		numNewMiniBlocks := 0
-		if len(miniBlock.TxHashes) == 0 {
-			numNewMiniBlocks = 1
-		}
-		numNewTxs := 1
-
-		isCrossShardScCallOrSpecialTx := receiverShardID != txs.shardCoordinator.SelfId() &&
-			(core.IsSmartContractAddress(tx.RcvAddr) || len(tx.RcvUserName) > 0)
-		if isCrossShardScCallOrSpecialTx {
-			if !firstCrossShardScCallOrSpecialTxFound {
-				numNewMiniBlocks++
-			}
-			numNewTxs += common.AdditionalScrForEachScCallOrSpecialTx
-		}
-
-		if isMaxBlockSizeReached(numNewMiniBlocks, numNewTxs) {
-			log.Debug("max txs accepted in one block is reached",
-				"num txs added", numTxsAdded,
-				"total txs", len(sortedTxs))
+	for _, wtx := range sortedTxs {
+		canAddTx, canContinue, tx := mbBuilder.checkAddTransaction(wtx)
```

Comment:
are both canAdd and canContinue needed? I think that returning just canContinue should suffice